File: Lung/ElBicho/trainModel.R
library( ANTsR )
library( ANTsRNet )
library( keras )
library( tensorflow )
library( ggplot2 )

keras::backend()$clear_session()

Sys.setenv( "CUDA_VISIBLE_DEVICES" = "2" )

baseDirectory <- '/home/ntustison/Data/Lung/'
# baseDirectory <- '/Users/ntustison/Data/HeliumLungStudies/DeepLearning/'
scriptsDirectory <- paste0( baseDirectory, 'Scripts/' )
source( paste0( scriptsDirectory, 'batchGenerator.R' ) )

templateSize <- c( 256L, 256L )
numberOfSlicesPerImage = 5
classes <- c( 0, 1, 2, 3, 4 )

################################################
#
#  Create the model and load weights
#
################################################

numberOfClassificationLabels <- length( classes )

imageModalities <- c( "Vent", "Mask" )
channelSize <- length( imageModalities )

unetModel <- createUnetModel2D( c( templateSize, channelSize ),
  numberOfOutputs = numberOfClassificationLabels, mode = 'classification',
  numberOfLayers = 4, numberOfFiltersAtBaseLayer = 32, dropoutRate = 0.0,
  convolutionKernelSize = c( 3, 3 ), deconvolutionKernelSize = c( 2, 2 ),
  weightDecay = 1e-5, addAttentionGating = TRUE )

weightsFileName <- paste0( scriptsDirectory, "/ventilationWeights.h5" )
if( file.exists( weightsFileName ) )
  {
  load_model_weights_hdf5( unetModel, filepath = weightsFileName )
  }

weighted_loss <- weighted_categorical_crossentropy( weights = c( 1, 10, 2, 1, 1 ) )

dice_loss <- multilabel_dice_coefficient( dimensionality = 2L, smoothingFactor = 0.1 )

# metric_multilabel_dice_coefficient <-
#  custom_metric( "multilabel_dice_coefficient",
#    multilabel_dice_coefficient )

#loss_dice <- function( y_true, y_pred ) {
#   -multilabel_dice_coefficient(y_true, y_pred )
#   }
#attr(loss_dice, "py_function_name") <- "multilabel_dice_coefficient"

unetModel %>% compile(
  optimizer = optimizer_adam(),
  loss = dice_loss, # weighted_loss, categorical_focal_loss( alpha = 0.25, gamma = 2.0 ),
  metrics = c( 'accuracy', metric_categorical_crossentropy ) )

################################################
#
#  Load the data
#
################################################

cat( "Loading data.\n" )

images <- c(
  Sys.glob( "/home/ntustison/Data/Lung/Ventilation/Images/*Ventilation.nii.gz" )
)

trainingImageFiles <- c()
trainingMaskFiles <- c()

pb <- txtProgressBar( min = 0, max = length( images ), style = 3 )
for( i in seq_len( length( images ) ) )
  {
  setTxtProgressBar( pb, i )

  image <- images[i]
  segmentation <- gsub( "Images", "Segmentations", image )
  segmentation <- gsub( "Ventilation.nii.gz", "Segmentation.nii.gz", segmentation )

  if( ! file.exists( image ) )
    {
    cat( "Image: ", image )  
    stop( "Image doesn't exist." )
    next
    }

  if( ! file.exists( segmentation ) )
    {
    cat( "Mask: ", segmentation )  
    stop( "Mask doesn't exist." )
    next
    }

  trainingImageFiles <- append( trainingImageFiles, image )
  trainingMaskFiles <- append( trainingMaskFiles, segmentation )
  }
cat( "\n" )

cat( "Total training image files: ", length( trainingImageFiles ), "\n" )

cat( "\nTraining\n\n" )


###
#
# Set up the training generator
#

batchSize <- 128L

# Split trainingData into "training" and "validation" componets for
# training the model.

numberOfData <- length( trainingImageFiles )
sampleIndices <- sample( numberOfData )

validationSplit <- floor( 0.8 * numberOfData )
trainingIndices <- sampleIndices[1:validationSplit]
numberOfTrainingData <- length( trainingIndices )
validationIndices <- sampleIndices[( validationSplit + 1 ):numberOfData]
numberOfValidationData <- length( validationIndices )

###
#
# Run training
#

track <- unetModel %>% fit_generator(
  generator = batchGenerator( batchSize = batchSize,
                              imageSize = templateSize,
                              images = trainingImageFiles[trainingIndices],
                              segmentations = trainingMaskFiles[trainingIndices],
                              labels = classes,
                              numberOfSlicesPerImage = numberOfSlicesPerImage,
                              doRandomContralateralFlips = TRUE,
                              doHistogramIntensityWarping = TRUE,
                              doDataAugmentation = FALSE
                            ),
  steps_per_epoch = 64L,
  epochs = 256L,
  validation_data = batchGenerator( batchSize = batchSize,
                                    imageSize = templateSize,
                                    images = trainingImageFiles[validationIndices],
                                    segmentations = trainingMaskFiles[validationIndices],
                                    labels = classes,
                                    numberOfSlicesPerImage = numberOfSlicesPerImage,
                                    doRandomContralateralFlips = TRUE,
                                    doHistogramIntensityWarping = TRUE,
                                    doDataAugmentation = FALSE
                                  ),
  validation_steps = 64L,
  callbacks = list(
    callback_model_checkpoint( weightsFileName,
      monitor = 'val_loss', save_best_only = TRUE, save_weights_only = TRUE,
      verbose = 1, mode = 'auto' ),
     callback_reduce_lr_on_plateau( monitor = 'val_loss', factor = 0.5,
       verbose = 1, patience = 10, mode = 'auto' ),
    callback_early_stopping( monitor = 'val_loss', min_delta = 0.0000001,
      patience = 20 )
  )
)

save_model_weights_hdf5( unetModel, filepath = weightsFileName )

File: Lung/ElBicho/train_model.py
import ants
import antspynet

import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
import glob

import random
import math

import tensorflow as tf
import tensorflow.keras as keras
import tensorflow.keras.backend as K

K.clear_session()
gpus = tf.config.experimental.list_physical_devices("GPU")
if len(gpus) > 0:
    tf.config.experimental.set_memory_growth(gpus[0], True)

base_directory = '/home/ntustison/Data/Lung/'
scripts_directory = base_directory + 'Scripts/'

from batch_generator import batch_generator

template_size = (256, 256)
number_of_slices_per_image = 5
classes = tuple(list(range(5)))

################################################
#
#  Create the model and load weights
#
################################################

number_of_classification_labels = len(classes)
image_modalities = ("Vent", "Mask")
channel_size = len(image_modalities)

unet_model = antspynet.create_unet_model_2d((*template_size, channel_size),
    number_of_outputs=number_of_classification_labels, mode="classification",
    number_of_layers=4, number_of_filters_at_base_layer=32, dropout_rate=0.0,
    convolution_kernel_size=(3, 3), deconvolution_kernel_size=(2, 2),
    weight_decay=1e-5, nn_unet_activation_style=False, add_attention_gating=True)

weights_filename = scripts_directory + "ventilationWeights.h5"
if os.path.exists(weights_filename):
    unet_model.load_weights(weights_filename)

unet_loss = antspynet.multilabel_dice_coefficient(dimensionality=2, smoothing_factor=0.1)
# unet_loss = antspynet.multilabel_surface_loss(dimensionality=2)

unet_model.compile(optimizer=keras.optimizers.Adam(),
                   loss=unet_loss,
                   metrics=['accuracy'])


################################################
#
#  Load the data
#
################################################

print("Loading data.")

images = glob.glob("/home/ntustison/Data/Lung/Ventilation/Images/*Ventilation.nii.gz")

training_image_files = list()
training_mask_files = list()

for i in range(len(images)):

    image = images[i]
    segmentation = image.replace("Images", "Segmentations")
    segmentation = segmentation.replace("Ventilation.nii.gz", "Segmentation.nii.gz")

    if not os.path.exists(image) or not os.path.exists(segmentation):
        continue

    training_image_files.append(image)
    training_mask_files.append(segmentation)

print("Total training image files: ", len(training_image_files))

print( "Training")

###
#
# Set up the training generator
#

batch_size = 128

# Split trainingData into "training" and "validation" componets for
# training the model.

number_of_data = len(training_image_files)
sample_indices = random.sample(range(number_of_data), number_of_data)

validation_split = math.floor(0.8 * number_of_data)

training_indices = sample_indices[:validation_split]
number_of_training_data = len(training_indices)

sampled_training_image_files = list()
sampled_training_mask_files = list()

for i in range(number_of_training_data):
    sampled_training_image_files.append(training_image_files[training_indices[i]])
    sampled_training_mask_files.append(training_mask_files[training_indices[i]])

validation_indices = sample_indices[validation_split:]
number_of_validation_data = len(validation_indices)

sampled_validation_image_files = list()
sampled_validation_mask_files = list()

for i in range(number_of_validation_data):
    sampled_validation_image_files.append(training_image_files[validation_indices[i]])
    sampled_validation_mask_files.append(training_mask_files[validation_indices[i]])


track = unet_model.fit_generator(
   generator=batch_generator(batch_size=batch_size,
                             image_size=template_size,
                             images=sampled_training_image_files,
                             segmentations=sampled_training_mask_files,
                             labels=classes,
                             number_of_slices_per_image=number_of_slices_per_image,
                             do_random_contralateral_flips=True,
                             do_histogram_intensity_warping=True,
                             do_add_noise=True,
                             do_data_augmentation=False
                            ),
    steps_per_epoch=64,
    epochs=256,
    validation_data=batch_generator(batch_size=batch_size,
                                    image_size=template_size,
                                    images=sampled_validation_image_files,
                                    segmentations=sampled_validation_mask_files,
                                    labels=classes,
                                    number_of_slices_per_image=number_of_slices_per_image,
                                    do_random_contralateral_flips=True,
                                    do_histogram_intensity_warping=True,
                                    do_add_noise=True,
                                    do_data_augmentation=False
                                   ),
    validation_steps=64,
    callbacks=[
        keras.callbacks.ModelCheckpoint(weights_filename, monitor='val_loss',
            save_best_only=True, save_weights_only=True, mode='auto', verbose=1),
        keras.callbacks.ReduceLROnPlateau(monitor='val_loss', factor=0.5,
           verbose=1, patience=10, mode='auto'),
        keras.callbacks.EarlyStopping(monitor='val_loss', min_delta=0.001,
           patience=20)
        ]
    )

unet_model.save_weights(weights_filename)



